src/registry/textfield/validationSchema.ts

```diff
@@ -7,18 +7,27 @@ const getValidationSchema: GetValidationSchema<TextFieldComponentSchema> = (
   componentDefinition,
   {validatePlugins}
 ) => {
-  const {key, validate = {}, multiple} = componentDefinition;
+  const {key, validate = {}, multiple, errors} = componentDefinition;
   const {required, maxLength, pattern, plugins = []} = validate;
 
-  let schema: z.ZodFirstPartySchemaTypes = z.string();
-  if (maxLength !== undefined) schema = schema.max(maxLength);
+  let schema: z.ZodFirstPartySchemaTypes = z.string(
+    errors?.required ? {required_error: errors.required} : undefined
+  );
+
+  if (maxLength !== undefined) {
+    schema = errors?.maxLength
+      ? schema.max(maxLength, {message: errors.maxLength})
+      : schema.max(maxLength);
+  }
```

Comment:
```suggestion
    schema = schema.max(maxLength, {message: errors?.maxLength});
  }
```
the most important part is `schema.max(...)` here, as it encodes the actual validation rule. having to duplicate that is error-prone, because an implementation change or refactor requires you to take into account all occurrences - that's bound to result in bugs.

Instead, the types allow for `{message: undefined}`, so we can use that :-) 

I see one possible problem where the key is present in errors and has an empty string value, but I'd classify that as a bug in the backend rather than the frontend code.